Task 3

faucct opened the pull request:
Инвентарь вызывается по кнопке `I`. В нём можно надевать и снимать вещи.
Также теперь игра сохраняется после каждого действия и при перезапуске продолжается с того же места. Чтобы начать новую игру нужно проиграть, выиграть или нажать кнопку `New game` в инвентаре.

AlexKaneRUS committed:
inventory and level system added

AlexKaneRUS committed:
added __pycache__ to .gitignore

faucct committed:
requirements.txt

faucct committed:
Fixed relative path in model/items/item

faucct committed:
WorldModel.terrain_to_world_model(terrain) should be static

faucct committed:
Mobs and Player should not reference world_graph

faucct committed:
WorldModel should initialize graphic representation

faucct committed:
Remove unused WorldModel::current_location

faucct committed:
GraphRepr should be a class

faucct committed:
WorldModel should not reference GraphRepr

faucct committed:
GraphRepr should be replaced with pygame.sprite.Group

faucct committed:
HasImage was extracted from HasCoordinates and Terrain

faucct committed:
Simplify HasCoordinates::set_coordinates

faucct committed:
CanMove should be separate from HasImage

faucct committed:
World dimensions configuration should be measured in tiles

faucct committed:
WorldModel should not reference pygame models

faucct committed:
Game should be saved

faucct committed:
Button position configuration is simplified

faucct committed:
Inventory should have a new game button

faucct committed:
Keyboard handler should move and toggle inventory commands instead of Game

AlexKaneRUS committed:
fixed problem with inventory reinitialize error

faucct committed:
Reading terrain from file

faucct committed:
PR fixes
Snake case
Constant names should be uppercase

ottergottaott commented on src/model/items/item.py:
```diff
@@ -0,0 +1,97 @@
+
+max_buff = 10
+
+offensive_item_prefixes = ['Sword', 'Gun', 'Cannon', 'Knife']
```
Стоит спрятать эти листы в соответствующий класс, а также дать им имена в upper case

faucct replied:
Сделано.

ottergottaott commented on src/model/world_model.py:
```diff
@@ -75,32 +87,50 @@ def move_logic(self, obj, next_move):
+            x, y = dir
+            return 0 <= y < len(map) and 0 <= x < len(map[y]) and map[y][x].is_passable()
+
+        return {
```
Использование dict для хранения двумерного массива — весьма сомнительная идея. Из dict, например, очень сложно вытащить длину одной из осей. 

Но метод стал явно читаемее, так что ок

ottergottaott commented on src/model/characters/has_inventory.py:
```diff
@@ -0,0 +1,156 @@
+            self.reinit(x, y, width, height, new_game_command)
+
+        class Command:
+            def __init__(self, j, inv, new_game_command):
```
А почему `Command` вложена в `Inventory`? И аналогично `Inventory` в `HasInventory`

faucct replied:
> А почему Command вложена в Inventory

Полагаю, что потому, что она относится к `Inventory`, но я переименовал её в `ItemCommand`.

> И аналогично Inventory в HasInventory

Это вынес.

ottergottaott commented on src/model/world_model.py:
```diff
@@ -75,32 +87,50 @@ def move_logic(self, obj, next_move):
+            with_inventory.pickup_item(item)
+
+        def collision_process(obj, other_obj):
+            if isinstance(obj, Player) and issubclass(other_obj.__class__,
```
Все еще не ясно зачем `issubclass`. Поясните или поменяйте, пожалуйста

AlexKaneRUS replied:
А мы это поменяли! Но в PR к Roguelike 2...

faucct replied:
Поменял и здесь.

ottergottaott commented on src/model/world_model.py:
```diff
@@ -2,71 +2,83 @@
-            self.world_graph[(mob.x, mob.y)].object = None
-            self.mobs.remove(mob)
-            self.graph_repr['mobs'].remove(mob)
+    def move_logic(self, obj, next_move, sprites):
```
И в соседнем же методе совсем какая-то беда: несколько вложенных методов, которые вызывают друг друга, а в одном методе даже есть еще один вложенный метод!
При этом еще и имена аргументов вложенных методов перекрывают имена аргументов внешних

Я прекрасно понимаю, что у вас скорее всего не было времени нормально порефакторить этот код, но постарайтесь исправить. Можно попробовать вынести отдельную сущность со всеми методами, разгрузив немного `WorldModel`

faucct replied:
С тем, что перекрытие имён – это плохо, я согласен: это я поправлю. Но локальные методы по-моему тут лучше подходят чем вынос этой логики в отдельный класс, который будет захватывать `WorldModel` со `sprites` и обращаться к полям `WorldModel`.

ottergottaott replied:
> Но локальные методы по-моему тут лучше подходят чем вынос этой логики в отдельный класс, который будет захватывать WorldModel со sprites и обращаться к полям WorldModel.

В Python есть методы и никто не запрещал передавать `WorldModel` еще куда-то :)

У меня все еще есть пара аргументов о том, почему это все-таки плохо, но я не очень намерен сейчас спорить (можно когда-нибудь потом и в кулуарах) и проблема не супер критичная, поэтому ок, пусть будет так

К слову, после исправлений стало лучше.

faucct replied:
> В Python есть методы и никто не запрещал передавать WorldModel еще куда-то :)

Я к тому, что если это будет отдельный класс, который будет знать про поля `WorldModel`, то он будет сильно сопряжён с ним и лучше уж оставить этот код тут, чем придумывать публичный интерфейс для этих полей.

ottergottaott reviewed (changes requested):
Функциональность, к сожалению, не могу проверить, т.к. у меня при запуске просто черный экран. Если такая проблема появляется только у меня, то давайте попросим @yurii-litvinov посмотреть, мб, проблема в моем Mac 

По коду +/- нормально, оставил комментарии по тому, что сильно не нравится

faucct commented:
> Функциональность, к сожалению, не могу проверить, т.к. у меня при запуске просто черный экран. Если такая проблема появляется только у меня, то давайте попросим @yurii-litvinov посмотреть, мб, проблема в моем Mac
> 
> По коду +/- нормально, оставил комментарии по тому, что сильно не нравится

Попробуйте на python-3.6 запустить. С остальными какие-то проблемы у pygame.

ottergottaott commented:
> Попробуйте на python-3.6 запустить. С остальными какие-то проблемы у pygame.

Получилось, по функциональности все ок

faucct committed:
PR fixes
No variable shadowing in WorldModel
Function isinstance should be used instead of issubclass wherever possible
Item prefixes and suffixes constants should belong to Item
Inventory should not be nested in HasInventory
Inventory.Command should be Inventory.ItemCommand

faucct committed:
Updated UML diagram

AlexKaneRUS commented:
@aleksZubakov 

Кажется, тут мы всё поправили, о чём ты писал!

Даже автосгенеренную UML-диаграмму приложили...

faucct commented:
> Даже автосгенеренную UML-диаграмму приложили...

Я довёл её до ума ещё отдельно.

vladislavneon committed:
Add displaying characteristics
